mirrord-layer/src/sockets.rs

```diff
@@ -533,17 +505,105 @@ unsafe extern "C" fn accept4_detour(
     }
 }
 
+fn fcntl(fcntl_fd: c_int, orig_fd: c_int, cmd: c_int) -> c_int {
+    match cmd {
+        libc::F_DUPFD | libc::F_DUPFD_CLOEXEC => {
+            let mut sockets = SOCKETS.lock().unwrap();
+            if let Some(socket) = sockets.get(&orig_fd) {
+                let fcntl_socket = socket.clone();
+                sockets.insert(fcntl_fd as RawFd, fcntl_socket);
+            }
+        }
+        _ => (),
+    }
+    fcntl_fd
+}
+
+unsafe extern "C" fn fcntl_detour(fd: c_int, cmd: c_int, arg: ...) -> c_int {
+    let fcntl_fd = libc::fcntl(fd, cmd, arg);
+    if fcntl_fd == -1 {
+        error!("fcntl failed");
+        return fcntl_fd;
+    }
+    fcntl(fcntl_fd, fd, cmd)
+}
+
+fn dup(fd: c_int) -> c_int {
+    let dup_fd = unsafe { libc::dup(fd) };
+    if dup_fd == -1 {
+        error!("dup failed");
+        return dup_fd;
+    }
+    let mut sockets = SOCKETS.lock().unwrap();
+    if let Some(socket) = sockets.get(&fd) {
+        let dup_socket = socket.clone();
+        sockets.insert(dup_fd as RawFd, dup_socket);
+    }
+    dup_fd
+}
+
+unsafe extern "C" fn dup_detour(fd: c_int) -> c_int {
+    dup(fd)
+}
+
+fn dup2(oldfd: c_int, newfd: c_int) -> c_int {
+    if oldfd == newfd {
+        return newfd;
+    }
+    let dup2_fd = unsafe { libc::dup2(oldfd, newfd) };
+    if dup2_fd == -1 {
+        error!("dup2 failed");
+        return dup2_fd;
+    }
+
+    let mut sockets = SOCKETS.lock().unwrap();
+    if let Some(socket) = sockets.get(&oldfd) {
+        let dup2_socket = socket.clone();
+        sockets.insert(dup2_fd as RawFd, dup2_socket);
+    }
+    dup2_fd
+}
+
+unsafe extern "C" fn dup2_detour(oldfd: c_int, newfd: c_int) -> c_int {
+    dup2(oldfd, newfd)
+}
+
+#[cfg(target_os = "linux")]
+fn dup3(oldfd: c_int, newfd: c_int, flags: c_int) -> c_int {
+    let dup3_fd = unsafe { libc::dup3(oldfd, newfd, flags) };
+    if dup3_fd == -1 {
```

Comment:
The implementations for fcntl, dup, dup2, and dup3 are pretty much the same - any way to extract them to a common function?